resteasy-jaxrs/src/main/java/org/jboss/resteasy/core/ServerResponseWriter.java

```diff
@@ -402,15 +403,22 @@ public static void commitHeaders(BuiltResponse jaxrsResponse, HttpResponse respo
             while (it.hasNext())
             {
                Object next = it.next();
+               String typeName = next.getClass().getTypeName();
                if (next instanceof NewCookie)
                {
                   NewCookie cookie = (NewCookie) next;
                   response.addNewCookie(cookie);
                   it.remove();
                }
+               else if ("java.lang.String".equals(typeName) || "char[]".equals(typeName))

```

Comment:
Actually I think that it is not a good idea to only allow `NewCookie `, `String `or `char[] `as valid Set-Cookies header values.
Any object  should be considered as valid as long as there is an instance of `javax.ws.rs.ext.RuntimeDelegate.HeaderDelegate` able to convert it to a String or using it `toString() ` method.
It seems that such classes already exist in the project such as `org.jboss.resteasy.plugins.delegates.CookieHeaderDelegate` wich handle `javax.ws.rs.core.Cookie` or `org.jboss.resteasy.plugins.delegates.NewCookieHeaderDelegate` wich handle `javax.ws.rs.core.NewCookie`.

The java doc about `javax.ws.rs.core.Response.ResponseBuilder.header(String, Object)` is:

> 
>            Add an arbitrary header.
>           
>            @param name  the name of the header
>            @param value the value of the header, the header will be serialized
>                         using a {@link javax.ws.rs.ext.RuntimeDelegate.HeaderDelegate} if
>                         one is available via {@link javax.ws.rs.ext.RuntimeDelegate#createHeaderDelegate(java.lang.Class)}
>                         for the class of {@code value} or using its {@code toString} method
>                         if a header delegate is not available. If {@code value} is {@code null}
>                         then all current headers of the same name will be removed.
>            @return the updated response builder.

Reply:
javax.ws.rs.ext.RuntimeDelegate.HeaderDelegate  should not be a concern.
The call to response.getOutputHeaders().add in actuality calls
HttpServletResponseHeaders.add which in turn calls
HttpServletResponseHeaders.addResponseHeaders line 93
This method gets the proper RuntimeDelegate.HeaderDelegate and executes it.